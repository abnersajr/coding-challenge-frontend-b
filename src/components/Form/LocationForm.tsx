import React, { useState } from 'react'

import LocationInput from '../Form/LocationInput'

const LocationForm = () => {
  return (
    <>
      <LocationInput
        formField={{
          label: 'Origin',
        }}
        textInput={{
          placeholder: 'Select Origin',
          suggestions: ['Quebec'],
        }}
      />
      <LocationInput
        formField={{
          label: 'Destination',
        }}
        textInput={{
          placeholder: 'Select Destination',
          suggestions: ['Montreal'],
        }}
      />
    </>
  )
}

export default LocationForm
