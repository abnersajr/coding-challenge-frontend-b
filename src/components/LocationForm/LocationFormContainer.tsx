import styled from 'styled-components'
import { Box } from 'grommet'

export const LocationFormContainer = styled.div`
  width: 100%;
  text-align: center;
`
export const LocationFormBox = styled(Box)`
  text-align: left;
  background: ${(props) => props.theme.colors.white};
  border-radius: ${(props) => props.theme.borderRadius};
  box-shadow: 0px 6px 12px rgb(32 65 90 / 10%), 0px 1px 4px rgb(32 65 90 / 10%);
  padding: 0;
  display: inline-flex;
  margin: ${(props) => props.theme.padding} auto;
  justify-content: center;
  > div {
    flex-basis: 100%;
  }
  @media screen and (min-width: 900px) {
    > div {
      flex-basis: auto;
    }
  }
`
