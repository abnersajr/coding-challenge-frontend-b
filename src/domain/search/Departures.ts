export type Departures = {
  origin_city_id: string
  destination_city_id: string
  cities: []
  locations: []
  operators: []
  departures: []
  complete: boolean
  status?: DeparturesStatus
}

export const NOT_INITIALIZED = 'NOT_INITIALIZED'
export const PENDING = 'PENDING'
export const INCOMPLETE = 'INCOMPLETE'
export const COMPLETE = 'COMPLETE'
export const REJECTED = 'REJECTED'

export type DeparturesStatus =
  | typeof NOT_INITIALIZED
  | typeof PENDING
  | typeof INCOMPLETE
  | typeof COMPLETE
  | typeof REJECTED

export const setCompleteOrIncomplete = (status: boolean) => {
  return status ? COMPLETE : INCOMPLETE
}
