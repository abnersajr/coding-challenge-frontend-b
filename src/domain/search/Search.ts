import { LocationDomain } from '../location'
import { LanguageDomain } from '../language'
import { CurrencyDomain } from '../currency'
export interface SearchData {
  origin: LocationDomain.Location
  destination: LocationDomain.Location
  outboundDate: string
  adult: number
  child: number
  senior: number
  senior_ages: string[]
  child_ages: string[]
  lang: LanguageDomain.Language
  currency: CurrencyDomain.Currency
}

export interface PathParams {
  origin: LocationDomain.Location['geohash']
  destination: LocationDomain.Location['geohash']
  outbound_date: string
}

export interface QueryParams {
  adult: number
  child: number
  senior: number
  senior_ages: string
  child_ages: string
  lang: LanguageDomain.Language
  currency: CurrencyDomain.Currency
}

export const CHILD = 'child'
export const ADULT = 'adult'
export const SENIOR = 'senior'
export const SENIOR_AGES = 'senior_ages'
export const CHILD_AGES = 'child_ages'

export type PassengerKeys = typeof CHILD | typeof ADULT | typeof SENIOR
export type PassengerAgeKeys = typeof CHILD_AGES | typeof SENIOR_AGES

export const hasMaxPassengers = (count: number) => {
  return count === 5
}

export class Search {
  private _path: PathParams
  private _query: QueryParams

  constructor({
    origin,
    destination,
    outboundDate,
    adult,
    child,
    senior,
    senior_ages,
    child_ages,
    lang,
    currency,
  }: SearchData) {
    this._path = {
      origin: origin.geohash,
      destination: destination.geohash,
      outbound_date: new Date(outboundDate).toISOString().split('T')[0],
    }
    this._query = {
      adult,
      child,
      senior,
      senior_ages: senior_ages.map((age) => parseInt(age)).join(','),
      child_ages: child_ages.join(','),
      lang,
      currency,
    }
  }

  static start(data: SearchData) {
    const search = new Search(data)
    return search
  }

  urlParams() {
    const { origin, destination, outbound_date } = this._path
    return `${origin}/${destination}/${outbound_date}`
  }

  queryParams() {
    return this._query
  }
}
