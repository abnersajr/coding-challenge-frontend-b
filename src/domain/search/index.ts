export * as SearchDomain from './Search'
