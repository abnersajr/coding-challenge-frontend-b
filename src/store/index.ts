import { configureStore } from '@reduxjs/toolkit'

import rootReducer from './rootReducer'

export type State = ReturnType<typeof rootReducer>

const configureAppStore = (preloadedState?: any) => {
  const store = configureStore({
    reducer: rootReducer,
    preloadedState,
    devTools: {
      name: 'Busbud Challenge',
    },
  })

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./rootReducer', () => store.replaceReducer(rootReducer))
  }

  return store
}

export default configureAppStore
