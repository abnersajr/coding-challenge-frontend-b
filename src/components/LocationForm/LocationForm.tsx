import React, { useState } from 'react'
import useInterval from 'use-interval'
import { IntlText } from '../Intl/IntlText'
import { FormSearch } from 'grommet-icons'
import styled from 'styled-components'

import { LocationFormContainer, LocationFormBox } from './LocationFormContainer'
import LocationInput from '../LocationInput/LocationInput'
import LocationSwitch from '../LocationSwitch/LocationSwitch'
import DateInput from '../DateInput/DateInput'
import { CityDomain } from '../../domain/city'
import { DateDomain } from '../../domain/language'
import { useSearch } from '../../store/search/hooks'
import { useDepartures } from '../../store/departures/hooks'
import { PassengerSelect } from '../PassengerSelect/PassengerSelect'
import { DatePassengerGroup } from '../DatePassengerGroup/DatePassengerGroup'

const SearchButton = styled.button`
  display: inline-block;
  background: ${(props) => props.theme.colors.blueLight};
  border: none;
  padding: 0px 25px;
  font-size: 18px;
  color: #fff;
  cursor: pointer;
  transition: background 250ms ease-in;
  width: 100%;
  border-radius: 4px;
  margin: 10px;

  @media screen and (min-width: 900px) {
    width: auto;
    margin: 0;
    border-radius: 0 4px 4px 0;
  }

  &:hover {
    background: ${(props) => props.theme.colors.yellow};
  }
`

const SearchButtonIcon = styled.span`
  display: inline-block;
  height: 24px;
  transform: translateY(5px);
`

const LocationForm = () => {
  const {
    switchPlaces,
    setPlace,
    setDate,
    getOrigin,
    getDestination,
    getOutboundDate,
  } = useSearch()

  const {
    isDeparturesSearchIncomplete,
    fetchDepartures,
    pollDepartures,
  } = useDepartures()

  useInterval(
    () => pollDepartures(),
    isDeparturesSearchIncomplete ? 3000 : null
  )

  const locationsSuggestions = CityDomain.getNames()
  return (
    <LocationFormContainer>
      <LocationFormBox direction="row" wrap={true}>
        <LocationInput
          formField={{
            label: <IntlText id="origin" />,
          }}
          textInput={{
            name: 'origin',
            value: getOrigin.name,
            placeholder: <IntlText id="select.origin" />,
            suggestions: locationsSuggestions,
            onChange: (event) =>
              setPlace({ field: 'origin', location: event.target.value }),
            onSelect: (event) => {
              event.target?.blur()
              setPlace({ field: 'origin', location: event.suggestion })
            },
          }}
        />
        <LocationSwitch onClick={() => switchPlaces()} />
        <LocationInput
          formField={{
            label: <IntlText id="destination" />,
          }}
          textInput={{
            name: 'destination',
            value: getDestination.name,
            placeholder: <IntlText id="select.destination" />,
            suggestions: locationsSuggestions,
            onChange: (event) =>
              setPlace({ field: 'destination', location: event.target.value }),
            onSelect: (event) => {
              event.target?.blur()
              setPlace({ field: 'destination', location: event.suggestion })
            },
          }}
        />
        <DatePassengerGroup direction="row" flex="grow">
          <DateInput
            value={getOutboundDate}
            onChange={(event) =>
              setDate(DateDomain.dateISO(new Date(event.value as string)))
            }
          />
          <PassengerSelect />
        </DatePassengerGroup>
        <SearchButton onClick={() => fetchDepartures()}>
          <SearchButtonIcon>
            <FormSearch color="#FFF" />
          </SearchButtonIcon>
          <IntlText id="search" />
        </SearchButton>
      </LocationFormBox>
    </LocationFormContainer>
  )
}

export default LocationForm
