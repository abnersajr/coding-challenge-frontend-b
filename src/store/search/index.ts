import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { formatISO } from 'date-fns'

import { DateDomain, LanguageDomain } from '../../domain/language'
import { CurrencyDomain } from '../../domain/currency'
import { SearchData } from '../../domain/search/Search'
import { setCurrency } from '../currency'
import { LocationDomain } from '../../domain/location'

type SearchState = {
  form: SearchData & {
    isLoading: boolean
  }
}

const initialState: SearchState = {
  form: {
    origin: LocationDomain.getLocationByName('Québec'),
    destination: LocationDomain.getLocationByName('Montréal'),
    outboundDate: DateDomain.todayString(),
    adult: 0,
    child: 0,
    senior: 0,
    lang: LanguageDomain.EN,
    currency: CurrencyDomain.USD,
    isLoading: false,
  },
}

type SetPlaceAction = {
  field: 'origin' | 'destination'
  location: LocationDomain.Location['name']
}
const setPlace = (
  state: SearchState,
  action: PayloadAction<SetPlaceAction>
) => {
  const { field, location } = action.payload
  state.form[field] = LocationDomain.getLocationByName(location)
}

const setDate = (state: SearchState, action: PayloadAction<string>) => {
  state.form.outboundDate = action.payload
}

const switchPlaces = (state: SearchState) => {
  const newOrigin = state.form.destination
  const newDestination = state.form.origin
  state.form.destination = newDestination
  state.form.origin = newOrigin
}

const currencySlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    setPlace,
    switchPlaces,
    setDate,
  },
  extraReducers: (builder) => {
    builder.addCase(
      setCurrency,
      (state, action: PayloadAction<CurrencyDomain.Currency>) => {
        state.form.currency = action.payload
      }
    )
  },
})

export const { reducer: searchReducer, actions } = currencySlice
