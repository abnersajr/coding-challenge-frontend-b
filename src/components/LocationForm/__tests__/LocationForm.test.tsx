import React from 'react'
import LocationForm from '../LocationForm'
import { render } from 'test/components'
import { screen, fireEvent } from '@testing-library/react'
import { initialState } from '../../../store/search'
describe('LocationForm', () => {
  it('renders correct', () => {
    render(<LocationForm />, {
      initialState: {
        search: {
          form: {
            ...initialState.form,
            outboundDate: new Date('2022-10-22T00:00Z').toISOString(),
          },
        },
      },
    })
    expect(screen.getByTestId('LOCATION.CONTAINER')).toMatchSnapshot()
  })
})
