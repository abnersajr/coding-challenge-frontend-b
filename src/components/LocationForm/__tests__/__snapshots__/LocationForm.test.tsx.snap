// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationForm renders correct 1`] = `
<div
  class="sc-bdfBwQ eNKbnG"
  data-testid="LOCATION.CONTAINER"
>
  <div
    class="StyledBox-sc-13pk1d4-0 gzJnOw sc-gsTCUz gMThGG"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 bGTiyL FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 jSBJfW"
      >
        Origin
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 dYebPD FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 kgYyVh"
            data-testid="INPUT.ORIGIN"
            name="origin"
            readonly=""
            value="Quebec"
          />
        </div>
      </div>
    </div>
    <button
      class="StyledButton-sc-323bzc-0 ebqXLG sc-dlfnbm cpUHiu"
      data-testid="PLACE_SWITCH"
      type="button"
    >
      <svg
        aria-label="Transaction"
        class="StyledIcon-ofa7kd-0 dEeaTZ"
        viewBox="0 0 24 24"
      >
        <path
          d="M2,7 L20,7 M16,2 L21,7 L16,12 M22,17 L4,17 M8,12 L3,17 L8,22"
          fill="none"
          stroke="#000"
          stroke-width="2"
        />
      </svg>
    </button>
    <div
      class="StyledBox-sc-13pk1d4-0 bGTiyL FormField__FormFieldBox-m9hood-0 jGEMXj"
    >
      <label
        class="StyledText-sc-1sadyjn-0 jSBJfW"
      >
        Destination
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 dYebPD FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 kgYyVh"
            data-testid="INPUT.DESTINATION"
            name="destination"
            readonly=""
            value="Montreal"
          />
        </div>
      </div>
    </div>
    <div
      class="StyledBox-sc-13pk1d4-0 gwCBYn sc-iqHYGH eLhfkV"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 jiuZyo FormField__FormFieldBox-m9hood-0 jGEMXj"
      >
        <label
          class="StyledText-sc-1sadyjn-0 jSBJfW"
        >
          Date
        </label>
        <div
          class="StyledBox-sc-13pk1d4-0 dYebPD FormField__FormFieldContentBox-m9hood-1"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dOerlP"
          >
            <div
              class="StyledMaskedInput__StyledIcon-sc-99vkfa-2 wgxqL"
            >
              <svg
                aria-label="Calendar"
                class="StyledIcon-ofa7kd-0 ebVQeN"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2,5 L22,5 L22,22 L2,22 L2,5 Z M18,5 L18,1 M6,5 L6,1 M2,10 L22,10"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </div>
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 jLezbT"
              data-testid="INPUT.DATE"
              name="date"
              placeholder="MM/dd/yyyy"
              readonly=""
              value="10/22/2022"
            />
          </div>
        </div>
      </div>
      <div
        class="StyledBox-sc-13pk1d4-0 bGTiyL FormField__FormFieldBox-m9hood-0 jGEMXj"
        data-testid="PASSENGER.TEXT"
      >
        <label
          class="StyledText-sc-1sadyjn-0 jSBJfW"
        >
          Passengers
        </label>
        <div
          class="StyledBox-sc-13pk1d4-0 dYebPD FormField__FormFieldContentBox-m9hood-1"
        >
          <button
            aria-label="Open Drop"
            class="StyledButton-sc-323bzc-0 hkpJCj Select__StyledSelectDropButton-sc-17idtfo-1 eYoNBZ"
            type="button"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 jmuHez"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 iAryDB"
              >
                <div
                  class="sc-kEjbxe kgPgIK"
                >
                  1 passenger
                </div>
              </div>
              <div
                class="StyledBox-sc-13pk1d4-0 eA-DBEo"
                style="min-width: auto;"
              >
                <svg
                  aria-label="FormDown"
                  class="StyledIcon-ofa7kd-0 kKTTyN"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    fill="none"
                    points="18 9 12 15 6 9"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
    <button
      class="sc-crrsfI kppKVf"
      data-testid="SEARCH.BUTTON"
    >
      <span
        class="sc-dQppl cmRHMh"
      >
        <svg
          aria-label="FormSearch"
          class="StyledIcon-ofa7kd-0 hXPYKN"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.8,13.8 L18,18 L13.8,13.8 Z M10.5,15 C12.9852814,15 15,12.9852814 15,10.5 C15,8.01471863 12.9852814,6 10.5,6 C8.01471863,6 6,8.01471863 6,10.5 C6,12.9852814 8.01471863,15 10.5,15 Z"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </span>
      Search
    </button>
  </div>
</div>
`;
